Docs: Add NHWC Batchnorm documentation (#3705)
ROCm/MIOpen: docs/how-to/use-nhwc-batchnorm-in-pytorch.rst

Supported configurations
=================================================

The following table shows the Batchnorm support for NHWC and NCHW with various data types and modes.
It also indicates which backend is used with and without the ``PYTORCH_MIOPEN_SUGGEST_NHWC_BATCHNORM``
environment variable enabled.

.. csv-table::
   :header: "Input data type","Memory format","Mode","Default backend","Backend with variable enabled"
   :widths: 20, 20, 20, 20, 25

   "``float32``","NCHW","1D/2D","MIOpen","MIOpen"
   "``float32``","NHWC","2D","native","MIOpen"
   "``float16``","NCHW","1D/2D mixed","MIOpen","MIOpen"
   "``float16``","NCHW","1D/2D not mixed","native","native"
   "``float16``","NHWC","2D mixed","native","MIOpen"
   "``float16``","NHWC","2D not mixed","native","native"
   "``bfloat16``","NCHW","1D/2D mixed","MIOpen (*)","MIOpen (*)"
   "``bfloat16``","NCHW","1D/2D not mixed","native","native"
   "``bfloat16``","NHWC","2D mixed","native","MIOpen"
   "``bfloat16``","NHWC","2D not mixed","native","native"
   "any","any","3D","native","native"

> (*) MIOpen is used with ROCm 6.5 and later. Otherwise, the native backend is used.

Reviewer: bfloat16 mixed batchnorn enabled 6.4 and later

In addition to Batchnorm, NHWC support is also available for convolution and other MIOpen features.

By default, NHWC Batchnorm support in MIOpen is disabled.
However, it can be enabled in a PyTorch environment using ROCm version 6.5 or later.
This configuration only supports 2D NHWC Batchnorm, with 3D NHWC Batchnorm planned for a future release.
1D Batchnorm is not applicable to the NHWC format.
To enable NHWC Batchnorm, use this command:

.. code:: shell

   PYTORCH_MIOPEN_SUGGEST_NHWC_BATCHNORM=1

.. note::

   In a future release, NHWC Batchnorm will be enabled by default and this command will be deprecated and
   removed.

PyTorch branch support
------------------------

Only the ``ROCm/pytorch`` PyTorch images support NHWC Batchnorm. ROCm version 6.5 or later is required.
The ``upstream`` images do not support this feature. The following PyTorch branches support the
NHWC Batchnorm feature:

*  `release/2.6 <https://github.com/ROCm/pytorch/tree/release/2.6>`_
*  `release/2.7 <https://github.com/ROCm/pytorch/tree/release/2.7>`_
*  `rocm6.4_internal_testing <https://github.com/ROCm/pytorch/tree/rocm6.4_internal_testing>`_

> *  `rocm6.5_internal_testing <https://github.com/ROCm/pytorch/tree/rocm6.5_internal_testing>`_

Reviewer 2: Please remove the internal_testing branches from this list.

Author: Done

and the `Intel® Extension for PyTorch GitHub <https://intel.github.io/intel-extension-for-pytorch/cpu/latest/tutorials/features/nhwc.html>`_.

Batchnorm
=================================================

Batchnorm (Batch Normalization) enables higher learning rates and reduces initialization overhead by
normalizing layer inputs. Ordinarily, the distribution of the inputs to each layer changes as the
parameters to the previous layer change. This makes it more difficult to train deep learning models
and leads to lower learning rates. With Batchnorm, normalization is part of the architecture
and is performed for each training batch.

For more information on Batchnorm, see `Batch Normalization: Accelerating Deep Network Training by Reducing Internal Covariate Shift <https://arxiv.org/abs/1502.03167>`_.

Enabling NHWC Batchnorm for MIOpen using PyTorch
=================================================

The PyTorch open-source tensor library provides support for using NHWC Batchnorm with MIOpen.
In addition to Batchnorm, NHWC support is also available for convolution and other MIOpen features.

By default, NHWC Batchnorm support in MIOpen is disabled.

> However, it can be enabled in a PyTorch environment using ROCm version 6.5 or later.

Reviewer 2: @amd-jnovotny We probably want to replace references to 6.5 with 7.0 for user-facing documentation

Author: Yes, I will scan for this throughout the document. I'm planning to review the entire document again after changes with an eye to issues like this.

This configuration only supports 2D NHWC Batchnorm, with 3D NHWC Batchnorm planned for a future release.
1D Batchnorm is not applicable to the NHWC format.
To enable NHWC Batchnorm, use this command:

.. code:: shell

   PYTORCH_MIOPEN_SUGGEST_NHWC_BATCHNORM=1

.. note::

   In a future release, NHWC Batchnorm will be enabled by default and this command will be deprecated and
   removed.

PyTorch branch support
------------------------

> Only the ``ROCm/pytorch`` PyTorch images support NHWC Batchnorm. ROCm version 6.5 or later is required.

Reviewer 2: @amd-jnovotny We probably want to replace references to 6.5 with 7.0 for user-facing documentation

Author: Good catch! Thanks. Will fix today.

> The ``upstream`` images do not support this feature. The following PyTorch branches support the

Reviewer 2: In general, this kind of statement is at risk of getting outdated soon after publishing, as we might upstream it by the time the documentation goes out for ROCm 7.0. To decouple the upstreaming timeline from the documentation timeline, we should either avoid making a statement about upstream availability, or choose a fixed point such as a recent PyTorch release. "PyTorch upstream" refers to `main` branch, which is a living animal.

Author: @jithunnair-amd That's a good point. It's probably best to avoid mentioning upstream availability and just state the PyTorch branches that do support the feature.

environment variable enabled.

.. note::

   Mixed mode means that the Batchnorm module has a different type than the inputs, for example,
   an input or gradient data type of ``FP16`` or ``BF16`` and a Batchnorm type of ``FP32``.
   If the Batchnorm module has the same data type as the inputs, for instance, an
   input or gradient data type of ``FP32`` and a Batchnorm module that is also ``FP32``, the mode is
   "not mixed".

.. csv-table::
   :header: "Input data type","Memory format","Mode","Default backend","Backend with variable enabled"
   :widths: 20, 20, 20, 20, 25

   "``float32``","NCHW","1D/2D","MIOpen","MIOpen"
   "``float32``","NHWC","2D","native","MIOpen"
   "``float16``","NCHW","1D/2D mixed","MIOpen","MIOpen"
   "``float16``","NCHW","1D/2D not mixed","native","native"
   "``float16``","NHWC","2D mixed","native","MIOpen"
   "``float16``","NHWC","2D not mixed","native","native"
   "``bfloat16``","NCHW","1D/2D mixed","MIOpen (*)","MIOpen (*)"
   "``bfloat16``","NCHW","1D/2D not mixed","native","native"
   "``bfloat16``","NHWC","2D mixed","native","MIOpen"
   "``bfloat16``","NHWC","2D not mixed","native","native"
   "any","any","3D","native","native"

> (*) MIOpen is used with ROCm 6.4 and later. Otherwise, the native backend is used.

Reviewer 2: @dnikolaev-amd What definitive statement can we make regarding upstream release/2.7 branch for this? Might be important to clarify to users if this feature is available in upstream release or not.

Reviewer: @amd-jnovotny NHWC batchnorm is enabled by default in release/2.7. 
And it can be switched to use native batchnorm with environment variable `PYTORCH_MIOPEN_SUGGEST_NHWC_BATCHNORM=0 `

Reviewer: @amd-jnovotny @jithunnair-amd we can extend this table for 3D also for all existing MIOpen cases

Author: @dnikolaev-amd To confirm, NHWC Batchnorm support in MIOpen is now enabled by default in release/2.7 and disabled by defaults in release/2.6 (but can be enabled with `PYTORCH_MIOPEN_SUGGEST_NHWC_BATCHNORM=1`)?

Reviewer: @amd-jnovotny 
2.6 - disabled by default. Can be enabled with PYTORCH_MIOPEN_SUGGEST_NHWC_BATCHNORM=1 (!!!)
2.7 - enabled by default. Can be disabled with PYTORCH_MIOPEN_SUGGEST_NHWC_BATCHNORM=0 (!!!)
@jithunnair-amd should we enable NHWC by default in 2.6?

Author: Thanks. I'll update this section and will get you to review after.

Author: @dnikolaev-amd To avoid having to duplicate the table for 2.6 and 2.7, I'll change the columns to be "Backend with NHWC Batchnorm disabled" and "Backend with NHWC Batchnorm enabled"

   MIOpen(HIP): Command [LogCmdBNorm] ./bin/MIOpenDriver bnorm -n 4 -c 8 -H 2 -W 2 -m 1 --forw 0 -b 1 -s 1 --layout NCHW
   MIOpen(HIP): Command [LogCmdBNorm] ./bin/MIOpenDriver bnorm -n 2 -c 8 -H 8 -W 1 -m 1 --forw 1 -b 0 -r 1 -s 1 --layout NHWC
   MIOpen(HIP): Command [LogCmdBNorm] ./bin/MIOpenDriver bnorm -n 2 -c 8 -H 8 -W 1 -m 1 --forw 0 -b 1 -s 1 --layout NHWC
   MIOpen(HIP): Command [LogCmdBNorm] ./bin/MIOpenDriver bnorm -n 2 -c 8 -H 8 -W 1 -m 1 --forw 1 -b 0 -r 1 -s 1 --layout NCHW
   MIOpen(HIP): Command [LogCmdBNorm] ./bin/MIOpenDriver bnorm -n 2 -c 8 -H 8 -W 1 -m 1 --forw 0 -b 1 -s 1 --layout NCHW

Each line corresponds to a different command or operation.
The ``./bin/MIOpenDriver`` string indicates that MIOpen was used for the operation.
The ``--layout`` parameter shows whether NHWC or NCHW was used, for example, ``--layout NHWC`` means the
NHWC memory format was used.

Running Batchnorm tests
=======================

Several test suites are available for Batchnorm. To test Batchnorm training using both NHWC and NCHW,
run the following command:

.. code:: shell

>    python test_nn.py -v -k test_batchnorm_train

Reviewer 2: @dnikolaev-amd Is the test name pattern changed after introduction of 3D batchnorms support in our release branches?

To test Batchnorm inference using both memory formats, use this command:

.. code:: shell

>    python test_nn.py -v -k test_batchnorm_inference

Reviewer 2: @dnikolaev-amd Is the test name pattern changed after introduction of 3D batchnorms support in our release branches?

Reviewer: @jithunnair-amd @amd-jnovotny 
test name pattern was changed with 3D in release/2.6 and 2.7

Now we have:

- test_batchnorm_3D_inference
- test_batchnorm_3D_train
- test_batchnorm_2D_inference
- test_batchnorm_3D_train

.. meta::
  :description: Using NHWC Batchnorm in PyTorch
  :keywords: MIOpen, ROCm, API, documentation, NHWC Batchnorm, PyTorch

************************************************************************************************

> Using NHWC Batchnorm in PyTorch

Reviewer 3: Just confirming that it's in "in" - not "with" or "on"

Author: I actually like "on" more because this feature sits on top of the base layer.

************************************************************************************************

This topic explains how to use NHWC Batchnorm for MIOpen operations in PyTorch. NHWC is

> a deep learning memory format that has certain performance advantages over traditional

Reviewer 3: ```suggestion
This topic explains how to use NHWC Batchnorm for MIOpen operations in PyTorch. NHWC is
a deep-learning memory format that has certain performance advantages over traditional
```

Reviewer 3: Not sure about this change.

memory formats.

> For information on installing and using PyTorch with ROCm, see :doc:`PyTorch on ROCm <rocm-install-on-linux:install/3rd-party/pytorch-install>`.

Reviewer 3: ```suggestion
For information about installing and using PyTorch with ROCm, see :doc:`PyTorch on ROCm <rocm-install-on-linux:install/3rd-party/pytorch-install>`.
```

For a list of the ROCm components and features that PyTorch supports, see :doc:`PyTorch compatibility <rocm:compatibility/ml-compatibility/pytorch-compatibility>`.
For more background on using PyTorch and ROCm for AI tasks, see
:doc:`Training a model with PyTorch for ROCm <rocm:how-to/rocm-for-ai/training/benchmark-docker/pytorch-training>`.

NHWC versus NCHW
=================================================

NHWC (also known as "Channels Last") and NCHW are two types of memory formats for deep learning. They describe how

> multidimensional arrays (nD) are translated to a linear (1-dimensional) memory address space.

Reviewer 3: ```suggestion
NHWC (also known as "Channels Last") and NCHW are two types of memory formats for deep learning. They describe how
multidimensional arrays (nD) are translated to a linear (one-dimensional) memory address space.
```

*  NCHW (Number of samples, channels, height, width): This is the default data layout in which channels
   are stored separately from one another. The height and width information is stored after
   the channels.
*  NHWC (Number of samples, height, width, channels): In this alternative format, channels are stored next
   to each other after the height and width information.

The performance of NHWC is better than NCHW and is close to that observed when using blocked memory format. NHWC is also

> easier to work with for common operations.

Reviewer 3: ```suggestion
The performance of NHWC is better than that of NCHW and is close to that observed when using a blocked memory format. NHWC is also
easier to work with for common operations.
```

For more information about these memory formats, see the
`PyTorch memory format documentation <https://pytorch.org/tutorials/intermediate/memory_format_tutorial.html>`_

> and the `Intel® Extension for PyTorch GitHub <https://intel.github.io/intel-extension-for-pytorch/cpu/latest/tutorials/features/nhwc.html>`_.

Reviewer 3: ```suggestion
and the `Intel Extension for PyTorch GitHub <https://intel.github.io/intel-extension-for-pytorch/cpu/latest/tutorials/features/nhwc.html>`_.
```

Reviewer 3: I don't think trademarks are used in links

> Batchnorm

Reviewer 3: Is it Batchnorm or BatchNorm?
Is this term already in the ROCm docs?
Here's an example from another site:
https://docs.nvidia.com/cuda/archive/12.8.1/cuequivariance/api/generated/cuequivariance_torch.layers.BatchNorm.html

Author: Settled on Batch Normalization as the official term.

=================================================

Batchnorm (Batch Normalization) enables higher learning rates and reduces initialization overhead by
normalizing layer inputs. Ordinarily, the distribution of the inputs to each layer changes as the
parameters to the previous layer change. This makes it more difficult to train deep learning models
and leads to lower learning rates. With Batchnorm, normalization is part of the architecture
and is performed for each training batch.

For more information on Batchnorm, see `Batch Normalization: Accelerating Deep Network Training by Reducing Internal Covariate Shift <https://arxiv.org/abs/1502.03167>`_.

Enabling or disabling NHWC Batchnorm for MIOpen using PyTorch
=============================================================

The PyTorch open-source tensor library provides support for using NHWC Batchnorm with MIOpen.
In addition to Batchnorm, NHWC support is also available for convolution and other MIOpen features.

> NHWC Batchnorm support in MIOpen can be used in a PyTorch environment using ROCm version 7.0 or later.

Reviewer 3: ```suggestion
NHWC Batchnorm support in MIOpen can be used in a PyTorch environment using ROCm 7.0 or later.
```

This configuration supports 2D and 3D NHWC Batchnorm. 1D Batchnorm is not applicable to the NHWC format.

PyTorch branch support
------------------------

The ``ROCm/pytorch`` PyTorch images support NHWC Batchnorm. ROCm version 7.0 or later is required.
The following PyTorch branches support the NHWC Batchnorm feature:

*  `release/2.6 <https://github.com/ROCm/pytorch/tree/release/2.6>`_
*  `release/2.7 <https://github.com/ROCm/pytorch/tree/release/2.7>`_

In the ``release/2.7`` PyTorch branch, NHWC Batchnorm support in MIOpen is enabled by default.
To use the native Batchnorm approach with this image, use this command:

.. code:: shell

   PYTORCH_MIOPEN_SUGGEST_NHWC_BATCHNORM=0

In the ``release/2.6`` PyTorch branch, NHWC Batchnorm support in MIOpen is disabled by default.
To enable NHWC Batchnorm for this image, use this command:

.. code:: shell

   PYTORCH_MIOPEN_SUGGEST_NHWC_BATCHNORM=1

> For information on installing and using PyTorch on ROCm, see :doc:`PyTorch on ROCm <rocm-install-on-linux:install/3rd-party/pytorch-install>`.

Reviewer 3: ```suggestion
For information about installing and using PyTorch on ROCm, see :doc:`PyTorch on ROCm <rocm-install-on-linux:install/3rd-party/pytorch-install>`.
```

   "``bfloat16``","NCHW","1D/2D/3D","not mixed","native","native"
   "``bfloat16``","NHWC","2D/3D","mixed","native","MIOpen"
   "``bfloat16``","NHWC","2D/3D","not mixed","native","native"

(*) MIOpen is used with ROCm 6.4 and later. Otherwise, the native backend is used.

Disabling MIOpen for Batchnorm in PyTorch
=================================================

In some situations, you might not want to use MIOpen as the backend for Batchnorm operations.
To disable the use of MIOpen with Batchnorm, add this code to your application.

.. code:: python

   inp = torch.randn(size, requires_grad=True)
   grad = torch.randn(size, requires_grad=False)
   mod = nn.BatchNorm2d(inp.size(1), device="cuda")

   with torch.backends.cudnn.flags(enabled=False): # this line disables MIOpen for the two lines below, native batchnorm will be used

      out = mod(inp)
      out.backward(grad)

Verifying NHWC Batchnorm use with MIOpen
=================================================

For some operations, it can be difficult to determine the backend and memory format that were used.
To verify whether MIOpen is being used and whether the memory format is NHWC or NCHW, run your program

> with the following environment variable:

Reviewer 3: ```suggestion
For some operations, it can be difficult to determine the backend and memory format used.
To verify whether MIOpen is being used and whether the memory format is NHWC or NCHW, run your program
with the following environment variable:
```
